text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,195 @@
+[motivation]: #motivation
+
+Rust is very promising for security-critical applications due to its safety guarantees, but there currently are gaps in the ecosystem that prevent it. One of them is the lack of any infrastructure for security updates.
+
+Linux distributions alert you if you're running a vulnerable software version and you can opt in to automatic security updates. Cargo not only has no automatic update infrastructure, it doesn't even know which libraries or library versions went into compiling a certain binary, so there's no way to check if your system is vulnerable or not.
+
+The primary use case for this information is cross-referencing versions of the dependencies against [RustSec advisory database](https://github.com/RustSec/advisory-db) and/or third-party databases such as [Common Vulnerabilities and Exposures](https://en.wikipedia.org/wiki/Common_Vulnerabilities_and_Exposures). This also enables use cases such as ensuring a fix in a depencency has been propagated across the entirety of your fleet or preventing binaries with unvetted dependencies from accidentally reaching a production environment - all with zero bookkeeping.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Every time an executable is compiled with Cargo, the dependency tree of the executable is recorded in the binary. This includes the names, versions, dependency kind (build or runtime), and origin kind (crates.io, git, local filesystem, custom registry). Development dependencies are not recorded, since they cannot affect the final binary. All filesystem paths and URLs are redacted to preserve privacy. The data is encoded in JSON and compressed with zlib to reduce its size.
+
+This data can be recovered using existing tools like `readelf` or Rust-specific tooling. It can be then used to create a Software Bill of Materials in a common format, or audit the dependency list for known vulnerabilities.
+
+WASM, asm.js and embedded platforms are exempt from this mechanism by default since they have very strict code size requirements. For those platforms we encourage you to use tooling that record the hash of every executable in a database and associates the hash with its Cargo.lock, compiler and LLVM version used for the build.
+
+A configuration option can be used to opt out of this behavior if it is not desired (e.g. when building [extremely minimal binaries](https://github.com/johnthagen/min-sized-rust)).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+The version information is encoded in an additional arbitrary section of the executable by Cargo. The exact mechanism varies depending on the executable format (ELF, Mach-O, PE, etc.). The section name is `.dep-v0` across all platforms. The section name must be changed if breaking changes are made to the format.
+
+The data is encoded in JSON which is compressed with Zlib. All arrays a sorted not to disrupt reproducible builds.
```

Comment:
Unless there's a specific established standard we need to be compatible with here, can we please use zstd for this? It's typically smaller, typically faster to compress for the same size, and almost always faster to decompress.

Reply:
This has been tested, and the gains from zstd compression are [negligible](https://github.com/rust-lang/rfcs/pull/2801#issuecomment-1665073360) even at extreme compression ratios.

The decompression speed does not really matter because the data is very small (under 50kb uncompressed, even in sprawling dependency trees). It is a negligible part of the parsing time.

Zstd does have drawbacks, however. For one, zstd does not have a Rust compressor implementation, and its Rust decompressor implementation is slow (3x slower than the reference C impl). This creates a highly non-trivial, memory-unsafe C dependency for all consumers of this format. 

So using zstd will complicate parsing and create far more attack surface for no real gain.

Reply:
FWIW I would love to have a pure-Rust zstd compressor and fast decompressor. I'm not willing to block on this, though. If a pure-Rust version becomes available in the future, perhaps dep-v1 can switch to it.

Reply:
Object uses the ruzstd crate for decompressing zstd compressed sections. Ruzstd is a pure rust zstd decompressor. According to it's readme it is about 1.4-3.5x slower than the reference implementation.

Reply:
@joshtriplett Using zstd instead of zlib doesn't provide much benefit at all:

> I used `rust-audit-info` to extract auditable data embedded in [`cargo-binstall-aarch64-apple-darwin`](https://github.com/cargo-bins/cargo-binstall/releases/download/v1.1.2/cargo-binstall-aarch64-apple-darwin.zip).
> 
> The decompressed data is 28.9KB, after compressed by `gzip -9` it's about 5KB.
> Compressing it using `zstd --ultra -22 --long` takes about 4.8KB, apparently I was wrong about suggestion of using zstd instead of zlib sinc `zstd` produce data set that is only 0.2KB smaller.
> 
> Given that `cargo-binstall` has roughly 318 dependencies, 5KB is reasonable since our pre-built artifacts are already quite large (9.6M after decompressed).

Originally posted at https://github.com/rust-lang/rfcs/pull/2801#issuecomment-1665073360